File: backend/model/torrent/torrent_entity.go
package torrent

import (
	"github.com/AlexKalinckovich/Cipher-Torrent-Client/backend/internal/shared/custom_errors/abstract_error_code"
	"time"
)

type TorrentStatus string

const (
	StatusIdle        TorrentStatus = "idle"
	StatusDownloading TorrentStatus = "downloading"
	StatusSeeding     TorrentStatus = "seeding"
	StatusPaused      TorrentStatus = "paused"
)

type TorrentEntity struct {
	InfoHash    []byte    `gorm:"primaryKey;type:binary(20)" json:"-"`
	InfoBytes   []byte    `gorm:"type:mediumblob" json:"-"`
	Name        string    `gorm:"type:varchar(255)" json:"name"`
	SizeBytes   int64     `json:"size_bytes"`
	PieceLength int       `json:"piece_length"`
	IsPrivate   bool      `json:"is_private"`
	StoragePath string    `gorm:"type:varchar(2048)" json:"storage_path"`
	AddedAt     time.Time `json:"added_at"`
}

const TorrentEntityCode abstract_error_code.ErrorCode = "TORRENT_NOT_FOUND"

func (e TorrentEntity) EntityCode() abstract_error_code.ErrorCode {
	return TorrentEntityCode
}

func (e TorrentEntity) EntityName() string {
	return "Torrent"
}

type UserTorrentEntity struct {
	UserID          int64         `gorm:"primaryKey" json:"user_id"`
	TorrentInfoHash []byte        `gorm:"primaryKey;type:binary(20)" json:"-"`
	Status          TorrentStatus `gorm:"type:varchar(20);default:idle" json:"status"`
	Progress        float32       `gorm:"type:decimal(5,2);default:0.00" json:"progress"`
}

const UserTorrentEntityCode abstract_error_code.ErrorCode = "USER_TORRENT_NOT_FOUND"

func (e UserTorrentEntity) EntityCode() abstract_error_code.ErrorCode {
	return UserTorrentEntityCode
}

func (e UserTorrentEntity) EntityName() string {
	return "UserTorrent"
}

File: frontend/src/types/model/Torrent/torrentEntity.ts
import type { TorrentFile } from './torrentFile';
import type {SignatureDTO} from "@/types/model/Torrent/torrentSignature.ts";

export type TorrentStatus = 'idle' | 'downloading' | 'seeding' | 'paused';

export interface TorrentEntity {
    name: string;
    size_bytes: number;
    piece_length: number;
    is_private: boolean;
    storage_path: string;
    added_at: string;
}

export interface TorrentDTO {
    info_hash: string;
    creator_public_key : string
    name: string;
    size_bytes: number;
    storage_path: string;
    status: TorrentStatus;
    progress: number;
    added_at: string;
    files: TorrentFile[];
    signatures: SignatureDTO[];
}